### PR TITLE
[RW-9101][RW-9104][RW-9103][risk=no]: Support list/clone/rename/delete/copy rmd files

### PR DESCRIPTION
**Pull request by yonghaoy**
Description:

- List RMD files is supported by checking rmd pattern macher
- Clone/Delete/Copy/Rename are supported without anychanges. 
- Get kernel type is not supported for rmd file
- Convert RMD file to HTLM is to be implemented. 
- Note, I can not test it because I can not create rmd files in gcs bucket and open with RStudio yet.  

---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have added explanatory comments where the logic is not obvious
- [ ] I have run and tested this change locally, and my testing process is described here
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later
- [ ] If this includes an API change, I have run the E2E tests on this change against my local server with [yarn test-local](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples) because this PR won't be covered by the CircleCI tests 
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer in [Slack](https://pmi-engteam.slack.com/archives/C02MWP2RN5P)
- [ ] If this change impacts deployment safety (e.g. removing/altering APIs which are in use) I have documented these in the description
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and updated the appropriate API consumers
  * AoU UI
  * [Perf tests](https://github.com/broadinstitute/mcnulty/blob/develop/src/test/scala/services/AoU.scala)
  * [Researcher Directory export](https://github.com/all-of-us/workbench/wiki/Researcher-Directory-(RDR-export))
  * Cron tasks - for Offline*Controllers
  * SumoLogic - for EgressAlert

### CODE REVIEW

File: api/src/main/java/org/pmiops/workbench/notebooks/NotebooksServiceImpl.java [hunk 1]
```diff
@@ -320,6 +324,11 @@ public String getReadOnlyHtml(
   @Override
   public String adminGetReadOnlyHtml(
       String workspaceNamespace, String workspaceName, String notebookName) {
+    if (!NotebookUtils.isJupyterNotebook(notebookName)) {
+      throw new NotImplementedException(
+          String.format("%s type of file is not implemented yet", notebookName));
```
Comment (evrii) [inline api/src/main/java/org/pmiops/workbench/notebooks/NotebooksServiceImpl.java:329]:
```suggestion
          String.format("%s is a type of file that is not yet supported", notebookName));
```

What do you think of this alternative error message?

File: api/src/main/java/org/pmiops/workbench/notebooks/NotebooksService.java [hunk 2]
```diff
@@ -8,13 +10,10 @@
 
 public interface NotebooksService {
 
-  String NOTEBOOKS_WORKSPACE_DIRECTORY = "notebooks";
-  String NOTEBOOK_EXTENSION = ".ipynb";
-
   static String withNotebookExtension(String notebookName) {
```
Comment (evrii) [inline api/src/main/java/org/pmiops/workbench/notebooks/NotebooksService.java:13]: It seems like clone, rename, and copy all rely on withNotebookExtension. It looks like this is only supporting files with a Jupyter extension. Does this need to be adjusted in order to support those functions for rmd files?

**Review by evrii — COMMENTED** [review]
This looks good to me, but I am unclear how this allows for list/clone/delete/copy rmd files. Is it because now that rmd files are considered

**Comment by yonghaoy** [comment]
> This looks good to me, but I am unclear how this allows for list/clone/delete/copy rmd files. Is it because now that rmd files are considered notebook files

The  notebookName in the request path is the full name like file1.rmd or file2.ipynb. And APIs just go go the corresponding GCS bucket dir, find the file then rename/copy/delete, etc. The file type doesn't matter as long as they are in the right dir.

File: api/src/main/java/org/pmiops/workbench/notebooks/NotebooksServiceImpl.java (hunk 1 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/notebooks/NotebooksServiceImpl.java:329]: Sounds good!

File: api/src/main/java/org/pmiops/workbench/notebooks/NotebooksService.java (hunk 2 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/notebooks/NotebooksService.java:13]: ah, good catch! We have this to make sure it works with/without file extension. 

- copy - does not pass extension!
- rename - pass the extension
- duplicate - pass the extension
- delete - pass the extension

File: api/src/main/java/org/pmiops/workbench/api/NotebooksController.java [hunk 3]
```diff
@@ -70,14 +72,22 @@ private FileDetail copyNotebookImpl(
...
+          isRMarkdownNotebook(fromNotebookName)
+              ? NotebooksService.withRMarkdownExtension(copyRequest.getNewName())
+              : NotebooksService.withJupyterNotebookExtension(copyRequest.getNewName());
+
       fileDetail =
           notebooksService.copyNotebook(
               fromWorkspaceNamespace,
               fromWorkspaceId,
-              NotebooksService.withNotebookExtension(fromNotebookName),
+              NotebooksService.withJupyterNotebookExtension(fromNotebookName),
```
Comment (evrii) [inline api/src/main/java/org/pmiops/workbench/api/NotebooksController.java:87]: Does your comment above (about removing withNotebookExtension) apply to this line too?

**Review by evrii — APPROVED** [review]
Looks great. Thanks for doing this.

File: api/src/main/java/org/pmiops/workbench/api/NotebooksController.java (hunk 3 above)
Comment (yonghaoy) [reply api/src/main/java/org/pmiops/workbench/api/NotebooksController.java:87]: yup. also copied the same comment over

File: api/src/main/java/org/pmiops/workbench/api/NotebooksController.java [hunk 4]
```diff
@@ -68,14 +72,23 @@ private FileDetail copyNotebookImpl(
...
+      // TODO(yonghao): Remove withNotebookExtension after UI start setting extension.
       fileDetail =
           notebooksService.copyNotebook(
               fromWorkspaceNamespace,
               fromWorkspaceId,
-              NotebooksService.withNotebookExtension(fromNotebookName),
+              NotebooksService.withJupyterNotebookExtension(fromNotebookName),
               copyRequest.getToWorkspaceNamespace(),
               copyRequest.getToWorkspaceName(),
-              NotebooksService.withNotebookExtension(copyRequest.getNewName()));
```
Comment (jmthibault79) [inline api/src/main/java/org/pmiops/workbench/api/NotebooksController.java:78]: @NehaBroad here is the bug

**Pull request merged**